x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics.event;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.collect.MapBuilder;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xcontent.json.JsonXContent;
+import org.elasticsearch.xpack.application.analytics.AnalyticsTemplateRegistry;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This class represents Analytics events object meant to be emitted to the event queue.
+ */
+public class AnalyticsEvent implements Writeable, ToXContentObject {
+
+    public static final ParseField TIMESTAMP_FIELD = new ParseField("@timestamp");
+    public static final ParseField EVENT_FIELD = new ParseField("event");
+    public static final ParseField EVENT_ACTION_FIELD = new ParseField("action");
+    public static final ParseField DATA_STREAM_FIELD = new ParseField("data_stream");
+    public static final ParseField DATA_STREAM_TYPE_FIELD = new ParseField("type");
+    public static final ParseField DATA_STREAM_NAMESPACE_FIELD = new ParseField("namespace");
+    public static final ParseField DATA_STREAM_DATASET_FIELD = new ParseField("dataset");
+
+    /**
+     * Analytics context. Used to carry information to parsers.
+     */
+    public interface Context {
+        long eventTime();
+
+        Type eventType();
+
+        String eventCollectionName();
+    }
+
+    /**
+     * Analytics event types.
+     */
+    public enum Type {
+        PAGE_VIEW("page_view"),
+        SEARCH("search"),
+        SEARCH_CLICK("search_click");
```

Comment:
> @jimczi Does it need to be specific to search, can it be click?

It is easier to have separate event type for each type of click as matter of validation.